File: ep03-polynomial-traj/python/polynomial_traj.py
"""3차·5차 다항식 관절 궤적 — 부드럽게 출발해 부드럽게 멈추는 점대점 이동.

Amatta Academy | 자격증 · 로봇소프트웨어개발기사 · 실기 EP.3

실행:  python polynomial_traj.py
출력:  cubic.csv, quintic.csv  (열: t, q, qd, qdd = 시간, 각도, 각속도, 각가속도)

예제 조건(영상과 동일): 시작 각도 0도 → 목표 각도 90도, 총 시간 2초
  - 3차: 계수 a2 = 3·90/2² = 67.5, a3 = -2·90/2³ = -22.5
         → 1초(중간)에서 각도 45도, 최고 각속도 초당 67.5도, 가속도 135 → -135 (출발·도착 순간 점프)
  - 5차: q = 90·(10τ³ - 15τ⁴ + 6τ⁵), τ = t/T
         → 1초에서 45도, 최고 각속도 약 초당 84.4도, 최고 각가속도 약 130 (가속도까지 0에서 시작·0으로 끝남)
"""
import csv


def cubic(q0, qf, T, dt=0.01):
    d = qf - q0
    a2, a3 = 3 * d / T**2, -2 * d / T**3
    rows, t = [], 0.0
    while t <= T + 1e-9:
        q = q0 + a2 * t**2 + a3 * t**3
        qd = 2 * a2 * t + 3 * a3 * t**2
        qdd = 2 * a2 + 6 * a3 * t
        rows.append((t, q, qd, qdd))
        t += dt
    return rows


def quintic(q0, qf, T, dt=0.01):
    d = qf - q0
    rows, t = [], 0.0
    while t <= T + 1e-9:
        s = t / T
        q = q0 + d * (10 * s**3 - 15 * s**4 + 6 * s**5)
        qd = d * (30 * s**2 - 60 * s**3 + 30 * s**4) / T
        qdd = d * (60 * s - 180 * s**2 + 120 * s**3) / T**2
        rows.append((t, q, qd, qdd))
        t += dt
    return rows


def save(rows, path):
    with open(path, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["t", "q", "qd", "qdd"])
        for t, q, qd, qdd in rows:
            w.writerow([f"{t:.3f}", f"{q:.4f}", f"{qd:.4f}", f"{qdd:.4f}"])


if __name__ == "__main__":
    Q0, QF, T = 0.0, 90.0, 2.0          # 시작 각도(도), 목표 각도(도), 총 시간(초)
    for name, fn in (("cubic", cubic), ("quintic", quintic)):
        rows = fn(Q0, QF, T)
        save(rows, f"{name}.csv")
        mid = min(rows, key=lambda r: abs(r[0] - T / 2))
        vmax = max(r[2] for r in rows)
        amax = max(abs(r[3]) for r in rows)
        print(f"{name:8s} 중간({mid[0]:.2f}초) 각도 = {mid[1]:.2f}도 | "
              f"최고 각속도 = {vmax:.3f} | 최고 각가속도 = {amax:.1f} | "
              f"끝 각도 = {rows[-1][1]:.2f}도 · 끝 속도 = {rows[-1][2]:.4f}")
    print("저장: cubic.csv, quintic.csv")

File: ep03-polynomial-traj/cpp/polynomial_traj.cpp
// 3차·5차 다항식 관절 궤적 — C++ 버전. cubic.csv, quintic.csv 를 만든다.
//
// Amatta Academy | 자격증 · 로봇소프트웨어개발기사 · 실기 EP.3
//
// 컴파일·실행 (시험장 Visual Studio 개발자 명령 프롬프트):
//     cl /EHsc polynomial_traj.cpp
//     polynomial_traj.exe
// 컴파일·실행 (g++ / MinGW):
//     g++ -std=c++11 polynomial_traj.cpp -o polynomial_traj
//     ./polynomial_traj          (Windows: polynomial_traj.exe)
//
// 예제 조건(영상과 동일): 0도 → 90도, 2초. 시간 간격 0.01초.
#include <cmath>
#include <cstdio>
#include <vector>

struct Row { double t, q, qd, qdd; };   // 시각, 각도, 각속도, 각가속도

// 3차: 계수 두 줄 → 각도 / 각속도 / 각가속도
std::vector<Row> cubic(double q0, double qf, double T, double dt = 0.01) {
    double d = qf - q0, a2 = 3 * d / (T * T), a3 = -2 * d / (T * T * T);
    std::vector<Row> rows;
    for (double t = 0.0; t <= T + 1e-9; t += dt)
        rows.push_back({t, q0 + a2 * t * t + a3 * t * t * t,
                        2 * a2 * t + 3 * a3 * t * t, 2 * a2 + 6 * a3 * t});
    return rows;
}

// 5차: 정규화 시간 s = t/T 로 10·15·6 공식
std::vector<Row> quintic(double q0, double qf, double T, double dt = 0.01) {
    double d = qf - q0;
    std::vector<Row> rows;
    for (double t = 0.0; t <= T + 1e-9; t += dt) {
        double s = t / T;
        rows.push_back({t, q0 + d * (10 * std::pow(s, 3) - 15 * std::pow(s, 4) + 6 * std::pow(s, 5)),
                        d * (30 * s * s - 60 * std::pow(s, 3) + 30 * std::pow(s, 4)) / T,
                        d * (60 * s - 180 * s * s + 120 * std::pow(s, 3)) / (T * T)});
    }
    return rows;
}

static void save(const std::vector<Row>& rows, const char* path) {
    std::FILE* f = std::fopen(path, "w");
    if (f == nullptr) return;
    std::fprintf(f, "t,q,qd,qdd\n");
    for (const Row& r : rows)
        std::fprintf(f, "%.3f,%.4f,%.4f,%.4f\n", r.t, r.q, r.qd, r.qdd);
    std::fclose(f);
}

int main() {
    const double Q0 = 0.0, QF = 90.0, T = 2.0;
    save(cubic(Q0, QF, T), "cubic.csv");
    save(quintic(Q0, QF, T), "quintic.csv");
    std::printf("saved: cubic.csv, quintic.csv\n");
    return 0;
}
